osu.Game/Database/StandardisedScoreMigrationTools.cs

```diff
@@ -264,9 +266,63 @@ public static long ConvertFromLegacyTotalScore(ScoreInfo score, LegacyBeatmapCon
+                    {
+                        return (long)Math.Round((
+                            0
+                            + 300000 * Math.Pow(score.Accuracy, 8)
+                            + bonusProportion) * modMultiplier);
+                    }
+
+                    // Assumption :
+                    // - sliders and slider-ticks are uniformly spread arround the beatmap
+                    //      thus we can ignore them without losing much precision (consider a map of hit-circles only !)
+                    // - the Ok/Meh hit results are uniformly spread in the score
+                    //      thus we can simplify and consider each hit result to be score.Accuracy without losing much precision
+                    // What is strippedV1/strippedV3 :
+                    // This is the ComboScore of v1/v3 were we remove all (map-)constant multipliers and accuracy multipliers (including hit results),
+                    // based on the previous assumptions. For Scorev1, this is basically the sum of squared combos (because without sliders: object_count == combo).
+                    double maxStrippedV1 = Math.Pow(maximumLegacyCombo, 2);
+                    double maxStrippedV3 = Math.Pow(maximumLegacyCombo, 1 + ScoreProcessor.COMBO_EXPONENT);
+
+                    double strippedV1 = maxStrippedV1 * comboProportion / score.Accuracy;
+
+                    double strippedV1FromMaxCombo = Math.Pow(score.MaxCombo, 2);
+                    double strippedV3FromMaxCombo = Math.Pow(score.MaxCombo, 1 + ScoreProcessor.COMBO_EXPONENT);
+
+                    // Compute approximate lower estimate scorev3 for that play
+                    // That is, a play were we made biggest amount of big combos (Repeat MaxCombo + 1 remaining big combo)
+                    // And didn't combo anything in the reminder of the map
+                    double possibleMaxComboRepeat = Math.Floor(strippedV1 / strippedV1FromMaxCombo);
+                    double strippedV1FromMaxComboRepeat = possibleMaxComboRepeat * strippedV1FromMaxCombo;
+                    double remainingStrippedV1 = strippedV1 - strippedV1FromMaxComboRepeat;
+                    double remainingCombo = Math.Sqrt(remainingStrippedV1);
+                    double remainingStrippedV3 = Math.Pow(remainingCombo, 1 + ScoreProcessor.COMBO_EXPONENT);
+
+                    double newLowerStrippedV3 = (possibleMaxComboRepeat * strippedV3FromMaxCombo) + remainingStrippedV3;
+
+                    // Compute approximate upper estimate scorev3 for that play
+                    // That is, a play were all combos were equal (except MaxCombo)
+                    remainingStrippedV1 = strippedV1 - strippedV1FromMaxCombo;
+                    double remainingComboObjects = maximumLegacyCombo - score.MaxCombo - score.Statistics[HitResult.Miss];
+                    double remainingAverageCombo = remainingComboObjects > 0 ? remainingStrippedV1 / remainingComboObjects : 0;
```

Comment:
I don't even. What is this? `remainingStrippedV1` is not raw combo. It's like combo squared. Is this score? Why doesn't it have "score" in the variable name, then?

Reply:
remainingStrippedV1 is sum of Combo squared, which is equivalent to the sum of group_of_object*combo_of_group. Dividing by the total amount of objects, we get the (weighted) average combo.

Reply:
Weighted how? Why weighted at all? Why not use the actual combo lengths, as computed by remaining combo object count by the number of misses or whatever?

Reply:
## Weighted average combo:
Let me give an example. Suppose we have the following score:
- 1 combo-streaks of 100 (our max combo)
- 1 combo-streak of 60
- 1 combo-streak of 50
- 1 combo-streak of 30
(Note: we can not know that the non-max combos are 60 and 30, we only have access to the total strippedV1)
strippedV1 = 100^2 + 60^2 + 50^2 + 30^2
remainingStrippedV1 = 60^2 + 50^2 + 30^2
remainingComboObjects = 140
remainingAverageCombo = (60^2 + 50^2 + 30^2)/140 = 60*(60/140) + 50*(50/140) + 30*(30/140) = 50

Here, X/140 are the weights (proportion of objects multiplied by those combos).

## Why not 1 + COMBO_EXPONENT

Actually, this was a simplification, but if this help understanding, you can see it like this:
```cs
double remainingCombos = (remainingComboObjects / remainingAverageCombo)
remainingStrippedV3 =  remainingCombos * Math.Pow(remainingAverageCombo, 1 + ScoreProcessor.COMBO_EXPONENT);
```
That is, the division by `remainingAverageCombo` cancels out the +1 exponent.
I guess I should not have made that simplification to make it easier to understand ?

Reply:
Right, okay, this clears it up a little, but I'm still not seeing what the "average" is doing and why it's weighted... I would understand something like

```csharp
double remainingAverageComboLength = (objectCount - maxCombo) / missCount;
double remainingComboPortionInStandardisedScore = (missCount - 1) * Math.Pow(remainingAverageComboLength, 1 + ScoreProcessor.COMBO_EXPONENT);
```

but the weighted average just has me completely confounded. Isn't it majorly overestimating the average combo length on the remainder of the map? Taking average over squares like that is going to skew the result on the larger side...

Reply:
missCount can not be used, because in scoreV1, breaking combo on slider-ticks doesn't cause a miss.

Yes we are slightly overestimating the final score here, but it's the best we can do AND remember that we are computing an over-estimate here, and we compensate with the lower-estimate computed in the other part.

Though, note that it is actually better to use a weighted average than the true average (if we even had access to it) and we are less overestimating that you might think.
Let me reuse the example above:
We have: `remainingStrippedV1 = 60^2 + 50^2 + 30^2`
and ideally we would want to obtain:
`idealRemainingStrippedNew = 60^1.5 + 50^1.5 + 30^1.5 ~= 982.6`
The but we only have access to the (weighted) average combo we computed is `50`, which gives:
`remainingStrippedNew = (140/50)*50^1.5 ~= 989.9`
Which is very close to what we wanted. In the ideal case (all remaining combos are equal) the results is perfect, and in realistic scenarios like that one we get something really close to the actual value.

If we had access to the true average combo (46.66) the result would actually be worse-off:
`trueAverageRemainingStrippedNew =  (140/50)*46.66^1.5 ~= 892.6`
What we would ideally want (but don't have), is actually to have the weighted average SQUARE ROOT combo:
`(60^1.5+50^1.5+*30^1.5)/140 = 7.01 = 49.2^0.5`
And thus we would be able to obtain the ideal result:
`(140/49.2) * 49.2^1.5 ~= 982.6 == idealRemainingStrippedNew`

Reply:
Intuitively, the reason the weighted average is still better than the true average is because in both formulas, the "big combos dominate" thus they should be given more weight in the equation.

That is if a >> b, then a**2 + b**2 ~= a**2 and a**1.5 + b**1.5 ~= a**1.5.

Reply:
Sorry but I still don't get it.

Let's formalise things a little. Let $x_1, x_2, \dots, x_n$ be real positive numbers indicating the lengths of non-highest combos in the score. Based on that - using your current terminology - we have

$$ \texttt{remainingStrippedV1} = \sum x_i^2 $$

$$ \texttt{remainingComboObjects} = \sum x_i $$

Thus

$$ \texttt{remainingAverageCombo} = \frac{\sum x_i^2}{\sum x_i} $$

which would be interpreted as the "average combo portion per hitobject" for the remainder of the score I guess. Now your `remainingCombos` is

$$ \texttt{remainingCombos} = \frac{\texttt{remainingComboObjects}}{\texttt{remainingAverageCombo}} = \frac{\sum x_i^2}{\left( \sum x_i \right)^2} $$

I have no idea how to interpret this number. It is just a number. And yes while when you do

$$ \texttt{remainingStrippedV3} = \texttt{remainingCombos} \cdot \texttt{remainingAverageCombo}^{1.5} $$

you get

$$ \frac{\sum x_i^2}{\left( \sum x_i \right)^2} \cdot \left( \frac{\sum x_i^2}{\sum x_i} \right)^{1.5} = \left( \sum x_i \sum x_i^2 \right)^{0.5} $$

which _is_ the same as `remainingComboObjects * Math.Pow(remainingAverageCombo, ScoreProcessor.COMBO_EXPONENT)`, but... I cannot interpret these transformations. It just looks like things being done on numbers again.

Please help.

Reply:
Ok, maybe it's actually better to think in terms of contributions per object as originally intended in the code. What we want is the sum of combo to power 1.5, which can be rewritten in terms of contribution per object as:

$$
\sum{x_i^{1.5}} = \sum{x_i} \cdot  \frac{\sum{x_i^{1.5}}}{\sum{x_i}}
$$

Where the left term is the amount of objects, and the right term is the average ComboScore contribution per object
As we can not compute the right term, we can approximate it with:

$$
\frac{\sum{x_i^{1.5}}}{\sum{x_i}} \approx{} \left( \frac{\sum{x_i^{2}}}{\sum{x_i}} \right) ^{0.5}
$$

That is: we take the square root of the average ScoreV1 ComboScore contribution per object. This approximation is guaranteed to be an over-estimation (because big combos are given more weight than they should have), but in cases where all $x_i$ are equal ($x_i = x$), the approximation is perfect:

$$
\frac{\sum{x_i^{1.5}}}{\sum{x_i}} = \frac{n \cdot x^{1.5}}{n \cdot  x} =x^{0.5}
$$

$$
\left( \frac{\sum{x_i^{2}}}{\sum{x_i}} \right) ^{0.5} = \left( \frac{n \cdot x^{2}}{n \cdot x} \right) ^{0.5} =  \left( x \right) ^{0.5}
$$

Reply:
> This approximation is guaranteed to be bigger or equal to what we want (because big combos are given a bit more weight than they should have)

This is _very_ handwavy and I can't seem to figure out a more rigorous proof of this but it does appear to check out in practice empirically.